clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2202,40 +2198,45 @@ static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
   ArgOS << ">";
 }
 
-static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
-                                           const PrintingPolicy &TypePolicy) {
-
-  QualType FullyQualifiedType = TypeName::getFullyQualifiedType(T, Ctx, true);
-
-  const CXXRecordDecl *RD = T->getAsCXXRecordDecl();
-
-  if (!RD)
-    return eraseAnonNamespace(FullyQualifiedType.getAsString(TypePolicy));
-
-  // If kernel name type is a template specialization with enum type
-  // template parameters, enumerators in name type string should be
-  // replaced  with their underlying value since the enum definition
-  // is not visible in integration header.
-  if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
-    SmallString<64> Buf;
-    llvm::raw_svector_ostream ArgOS(Buf);
-
-    // Print the qualifiers for the type.
-    FullyQualifiedType.getQualifiers().print(ArgOS, TypePolicy,
+static std::string printRecordType(QualType T, const CXXRecordDecl *RD,
+                                   const PrintingPolicy &TypePolicy) {
+  SmallString<64> Buf;
+  llvm::raw_svector_ostream OS(Buf);
+  T.getCanonicalType().getQualifiers().print(OS, TypePolicy,
                                              /*appendSpaceIfNotEmpty*/ true);
+  if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
 
     // Print template class name
-    TSD->printQualifiedName(ArgOS, TypePolicy, /*WithGlobalNsPrefix*/ true);
+    TSD->printQualifiedName(OS, TypePolicy, /*WithGlobalNsPrefix*/ true);
 
     // Print template arguments substituting enumerators
     ASTContext &Ctx = RD->getASTContext();
     const TemplateArgumentList &Args = TSD->getTemplateArgs();
-    printTemplateArguments(Ctx, ArgOS, Args.asArray(), TypePolicy);
+    printTemplateArguments(Ctx, OS, Args.asArray(), TypePolicy);
+
+    return eraseAnonNamespace(OS.str().str());
+  }
+  if (RD->getDeclContext()->isFunctionOrMethod())
+    return eraseAnonNamespace(T.getCanonicalType().getAsString(TypePolicy));
+  if (dyn_cast<NamespaceDecl>(RD->getDeclContext()) &&
+      dyn_cast<NamespaceDecl>(RD->getDeclContext())->isAnonymousNamespace())
```

Comment:
You could avoid the second dynamic cast by storing the result of the first one.

Reply:
Yeah, you should also only be dyn_casting something if you're going to use the result, otherwise use isa.

Reply:
Also, odd to do this with an 'if/else' rather than just putting the condition in the call.

Reply:
With the new version I have I think responded to Mariya and Prem's review. Not sure what you mean Erich? Do you mean the caller of printRecordType should be called with the Namespace? Or do you mean the call of printQualified?

Reply:
I mean, it seems that the body of these two differs only by 'true' or 'false'.  Instead of the if/else, do:

    RD->printQualifiedName(OS, TypePolicy, NS && NS->isAnonymousNamespace());